### DISCUSSION TITLE
Tool-calling does not work consistently (compared to ChatGPT Actions)

### DISCUSSION THREAD

**Discussion by mattia-intellilab** [open-webui/open-webui#8172] [General]
# Bug Report

## Installation Method

We used docker in order to install openwebui on a Kubernetes cluster. 

## Environment

- **Open WebUI Version:** v0.4.6

- **Operating System:** Windows 11
- **Browser (if applicable):** Google Chrome

**Confirmation:**

- [x ] I have read and followed all the instructions provided in the README.md.
- [ ] I am on the latest version of both Open WebUI and Ollama.
- [ ] I have included the browser console logs.
- [ ] I have included the Docker container logs.
- [ ] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:

When activating the chat_to_make Tool. 
The Model should use it in order to send the chat-body to a make.com webhook. 

![image](https://github.com/user-attachments/assets/eb043da7-d216-41bc-bf03-cf6870392865)

## Actual Behavior:

2/10 times the tool gets executed correctly and the data gets to the webhook. 

Example where it worked: 
![image](https://github.com/user-attachments/assets/dd617d5f-205c-4e62-bd5d-f26722ea5ad7)
![image](https://github.com/user-attachments/assets/f0d2912d-369b-4411-9c5e-f9e0c4b1392d)

Now this time, with a bigger chat-context, the tool does not get called and there is now new data received by the webhook. 
![image](https://github.com/user-attachments/assets/f1b7aa72-dfe9-4700-9107-7cf59ab151c0)

Same here:
![image](https://github.com/user-attachments/assets/e60c9f44-93b0-422e-90f9-f1175990efba)

## Description

This would be the tool-code:

`import os
import requests
from typing import Optional, Callable, Awaitable
from pydantic import BaseModel, Field

class Tools:
    class Valves(BaseModel):
        webhook_url: str = Field(
            default="",
            description="The URL of the make.com webhook to send chat-data to.",
        )

    def __init__(self):
        self.valves = self.Valves()
        pass

    def send_chat(
        self,
        body: dict,
    ) -> str:
        """
        Send the chat-history to make.com using a webhook when chat_to_make is prompted.
        :param body: contains all messages from the user and the assistant.
        :return: None
        """
        messages = body.get("messages", [])
        if not messages:
            return "No messages found in the request body."

        user_messages = [
            message["content"] for message in messages if message["role"] == "user"
        ]

        assistant_messages = [
            message["content"] for message in messages if message["role"] == "assistant"
        ]

        webhook = self.valves.webhook_url
        if not webhook:
            return "Let the user know webhook URL was not provided. Please configure the webhook URL."

        payload = {
            "content": f"User Messages: {user_messages} Assistant Messages:{assistant_messages}  - Sent from Open WebUI"
        }
        response = requests.post(webhook, json=payload)

        if response.status_code == 200:
            return "Message successfully sent, Let the user know the message has been sent."
        else:
            return f"Failed to send message. HTTP Status Code: {response.status_code},  Let the user know there were some issues."
`

## Reproduction Details

**Steps to Reproduce:**

1. Setup a new tool using the code above
2. Activate the tool in a new chat with gpt4o
3. Use a sample prompt in order to populate the chat-body
4. Tell the model to use the tool "chat_to_make" in order to send chat-body to make.com webhook
5. check make.com webhook to see if the data has arrived

Any ideas how to fix this?

**Comment by lowlyocean** [comment]
Open WebUI currently uses its own prompt-based approach (agnostic to the specific model) to call tools. I've seen many report this to work well enough with large commercial models like ChatGPT, but not good for smaller open source models running locally. You may want to follow https://github.com/open-webui/open-webui/pull/7021 , which is about having Open WebUI use model-specific ("native") tool calling.

**Reply by mattia-intellilab** [reply]
I am actually using OpenAI GPT-4o. Thats what puzzlez me. 
When using a Tool like https://openwebui.com/t/abhiactually/perplexity there is no issue, calling the tool. 
However with my own tools, the feature feels rather unreliable.

**Reply by mattia-intellilab** [reply]
Would Function-Calling be a Solution here?

**Reply by marcelsamyn** [reply]
I want to have native function calling in Open-WebUI so have started looking into it.

The workflow right now seems to be like this:

- User presses "Send message"
- Message thread gets reformatted into a prompt like this: "Here's your tools: ..., here's the chat history: ..., here's the question you answer. Respond in JSON please."
- Model (hopefully) responds with JSON, which runs the tool and gives it back to the model
- Final answer gets appended to the message thread

There's a couple of non-ideal properties here:

- Function calls are not part of the message thread, so the model doesn't know which tools it called later on. I think this is why you get the weird response of "I can't send it" while it actually sent the response—the model is generating that reply only using the *output of the tool*, it doesn't see the available tools anymore.
- There's no "interplay" between function calls and the model; ChatGPT can do multiple calls, react to their output, try again etc.

Right now I've created a pipe that uses native function calling so you can try that out and see if it works better: https://openwebui.com/f/marcelsamyn/native_tool_calling_pipe

Hope I'll have some time soon to PR this functionality into the app!
